### PR TITLE
[BUGFIX]: Ensure empty response for hasMany relationship is properly reflected on client

### PR DESCRIPTION
**Pull request by snewcomer**
close https://github.com/emberjs/data/issues/7318
ref https://github.com/emberjs/data/issues/5204

- Notes and links in each commit when removing from unused methods.
- flushCanonicalLater to sync local state with "persisted" state.

https://discord.com/channels/480462759797063690/486549196837486592/748932008360542480
https://github.com/amk221/-ember-data-remove-has-many/blob/master/tests/unit/controllers/application-test.js#L53

### CODE REVIEW

File: packages/record-data/addon/-private/relationships/state/has-many.ts [hunk 1]
```diff
@@ -162,25 +170,9 @@ export default class ManyRelationship extends Relationship {
...
-      }
 
-      this.canonicalMembers.add(recordData);
-      this.members.add(recordData);
-      this.setupInverseRelationship(recordData);
-    }
-
-    this.canonicalState = this.canonicalMembers.toArray();
+    // sync with local state
+    this.flushCanonicalLater();
```
Comment (snewcomer) [inline packages/record-data/addon/-private/relationships/state/has-many.ts:175]: This was the main change - at the end of `computeChanges`, flush to sync with local state

File: packages/record-data/addon/-private/relationships/state/has-many.ts [hunk 2]
```diff
@@ -225,7 +217,7 @@ export default class ManyRelationship extends Relationship {
     return payload;
   }
 
-  updateData(data, initial) {
+  updateData(data) {
```
Comment (snewcomer) [inline packages/record-data/addon/-private/relationships/state/has-many.ts:220]: `initial` is never an argument passed

File: packages/record-data/addon/-private/relationships/state/relationship.ts [hunk 3]
```diff
@@ -620,7 +610,7 @@ export default class Relationship {
    `push` use `updateMeta`, `updateData` and `updateLink` to update the state
    of the relationship.
    */
-  push(payload: JsonApiRelationship, initial?: boolean) {
+  push(payload: JsonApiRelationship) {
```
Comment (snewcomer) [inline packages/record-data/addon/-private/relationships/state/relationship.ts:613]: Same here - never passed as an argument.

This seems to be accomplished with `addRecordData` && `addCanonicalRecordData`

Comment (snewcomer) [reply packages/record-data/addon/-private/relationships/state/relationship.ts:613]: https://github.com/emberjs/data/blob/master/packages/record-data/addon/-private/record-data.ts#L223

A specific use case to show `initial` is not a used argument for `push`

File: packages/record-data/addon/-private/relationships/state/belongs-to.ts [hunk 4]
```diff
@@ -50,20 +50,6 @@ export default class BelongsToRelationship extends Relationship {
...
-    }
-
-    // When we initialize a belongsTo relationship, we want to avoid work like
-    // notifying our internalModel that we've "changed" and excessive thrash on
-    // setting up inverse relationships
-    this.canonicalMembers.add(recordData);
-    this.members.add(recordData);
-    this.inverseRecordData = this.canonicalState = recordData;
-    this.setupInverseRelationship(recordData);
-  }
```
Comment (snewcomer) [inline packages/record-data/addon/-private/relationships/state/belongs-to.ts:65]: Without `initial` to `push`, this isn't needed

File: packages/record-data/addon/-private/relationships/state/relationship.ts [hunk 5]
```diff
@@ -331,16 +331,6 @@ export default class Relationship {
     });
   }
 
-  addCanonicalRecordDatas(recordDatas: RelationshipRecordData[], idx: number) {
```
Comment (snewcomer) [inline packages/record-data/addon/-private/relationships/state/relationship.ts:334]: Searching across the codebase, this isn't used anywhere

**Comment by amk221** [comment]
Got some code I'd love to ship but need this to go in first :D

**Comment by snewcomer** [comment]
@amk221 I'll see if we can discuss in the first e-d meeting of the year!

**Comment by amk221** [comment]
Hi all, really keen to get this fix in. Itching to release some new stuff

**Comment by amk221** [comment]
👋 Do you know of any temporary work arounds I can use in the meantime?

I've tried the following:

* Reloading the record 
* Pushing a new hasMany payload to the record 
* Unloading the record (works, but not a viable solution, because this unloads all other relationships too)

So, in a last ditch effort to ship, I've gone with this:

```javascript
export default class UserSerializer extends BaseSerializer {
  attrs = {
    roles: {
      embedded: 'always'
    }
  };

  normalizeUpdateRecordResponse(store, primaryModelClass, payload) {
    // Horrible hack
    if (payload.user.roles.length === 0) {
      window.location.reload();
    }

    return super.normalizeUpdateRecordResponse(...arguments);
  }
}
```

**Comment by runspired** [comment]
@amk221 this is likely fixed on master already, mind PR'ing just the test?

**Pull request closed without merging**